libazureinit/src/provision/mod.rs

```diff
@@ -23,156 +24,120 @@ use crate::User;
 }
 
 impl Provision {
-    pub fn new(hostname: impl Into<String>, user: User) -> Self {
+    pub fn new(
+        hostname: impl Into<String>,
+        user: User,
+        config: Config,
+    ) -> Self {
         Self {
             hostname: hostname.into(),
             user,
-            hostname_backends: None,
-            user_backends: None,
-            password_backends: None,
+            config,
         }
     }
 
-    /// Specify the ways to set the virtual machine's hostname.
-    ///
-    /// By default, all known methods will be attempted. Use this function to
-    /// restrict which methods are attempted. These will be attempted in the
-    /// order provided until one succeeds.
-    pub fn hostname_provisioners(
-        mut self,
-        backends: impl Into<Vec<hostname::Provisioner>>,
-    ) -> Self {
-        self.hostname_backends = Some(backends.into());
-        self
-    }
-
-    /// Specify the ways to create a user in the virtual machine
-    ///
-    /// By default, all known methods will be attempted. Use this function to
-    /// restrict which methods are attempted. These will be attempted in the
-    /// order provided until one succeeds.
-    pub fn user_provisioners(
-        mut self,
-        backends: impl Into<Vec<user::Provisioner>>,
-    ) -> Self {
-        self.user_backends = Some(backends.into());
-        self
-    }
-
-    /// Specify the ways to set a users password.
-    ///
-    /// By default, all known methods will be attempted. Use this function to
-    /// restrict which methods are attempted. These will be attempted in the
-    /// order provided until one succeeds. Only relevant if a password has been
-    /// provided with the [`User`].
-    pub fn password_provisioners(
-        mut self,
-        backend: impl Into<Vec<password::Provisioner>>,
-    ) -> Self {
-        self.password_backends = Some(backend.into());
-        self
-    }
-
-    /// Provision the host.
-    ///
     /// Provisioning can fail if the host lacks the necessary tools. For example,
-    /// if there is no `useradd` command on the system's `PATH`, or if the command
+    /// if there is no useradd command on the system's PATH, or if the command
     /// returns an error, this will return an error. It does not attempt to undo
     /// partial provisioning.
     #[instrument(skip_all)]
     pub fn provision(self) -> Result<(), Error> {
-        self.user_backends
-            .unwrap_or_else(|| user::Provisioner::iter().collect())
+        self.config
+            .hostname_provisioners
+            .backends
+            .iter()
+            .find_map(|backend| match backend {
+                HostnameProvisioner::Hostnamectl => {
+                    hostname::Provisioner::Hostnamectl.set(&self.hostname).ok()
+                }
+                #[cfg(test)]
+                HostnameProvisioner::FakeHostnamectl => Some(()),
+            })
+            .ok_or_else(|| Error::NoHostnameProvisioner)?;
```

Comment:
This is super helpful, thank you! I'll go ahead and make the change just because I support any changes to making it better, no matter how small, but thank you again for pointing that out!

Reply:
@jeremycline This comment definitely sparked my interest some of these subtle differences between Rust constructs. It seems to me that other than slightly better readability, there's not much advantage to using `ok_or`.
I shamelessly took this [sample code ](https://medium.com/@techhara/rust-tip-and-trick-2-5a92641191f6#:~:text=The%20difference%20between%20Result%3A%3A,argument%2C%20rather%20than%20a%20value.) and modified it to use the same `thiserror` enum that we're using and ran a little experiment comparing `ok_or(Error::NoHostnameProvisioner.to_string())` vs `ok_or_else(|| Error::NoHostnameProvisioner.to_string()). `ok_or_else` outperformed `ok_or` by 30%-40% in release build and about 10% in debug build. 

This level of optimization doesn't really matter for azure-init (since we have many expensive operations like calling into IMDS, wireserver...) but I why wouldn't it be ok using `or_else` constructs everywhere just to make it simple?